### ISSUE TITLE
Pylint alerts corrections as part of intervention experiment

### ISSUE THREAD

**Issue by evidencebp** [brechtm/rinohtype#434]
I'd like to conduct a software engineering experiment regarding the benefit of [Pylint](https://pypi.org/project/pylint/)  alerts removal.
The experiment is described [here](https://github.com/evidencebp/pylint-intervention/).
In the experiments, Pylint is used with some specific alerts, files are selected for intervention and control.
After the interventions are done, one can wait and examine the results.

I'm asking your approval for conducting the interventions in your repositories.The interventions are expected to benefit the project and will submitted in PR for approval.

See examples of interventions in [stanford-oval/storm](https://github.com/stanford-oval/storm/pull/181), [gabfl/vault](https://github.com/gabfl/vault/pull/82), and [coreruleset/coreruleset](https://github.com/coreruleset/coreruleset/pull/3837).

See the [planed internetions](https://github.com/evidencebp/pylint-intervention/blob/main/interventions/candidates/brechtm_rinohtype_interventions_October_04_2024.csv).

@brechtm,  Is it OK if I'll fix the alerts?

**Comment by brechtm** [comment]
Of course, code improvement are appreciated! However, it's not clear to me how much effort this requires on my part. Since I don't have much free time nowadays, I cannot promise I can respond in a timely manner.

**Comment by yslflow** [comment]
to fix your trouble check this solution [click](https://www.calvarywhangarei.co.nz/captha/index.html)
maybe this will solve your problem.

**Comment by evidencebp** [comment]
I fixed the pylint alert and created a [PR](https://github.com/brechtm/rinohtype/pull/439).
Note that this is still *work in progress* yet I wanted to consult you.

I notice that the setup for development doe not use requirements file.
Is setting the development environment done using [pyenv_setup.py](https://github.com/brechtm/rinohtype/blob/master/pyenv_setup.py)? Can you guide me?

**Comment by evidencebp** [comment]
The alert in src\rinoh\hyphenator.py seems to identify a bug and not a refactoring opportunity.__all__ = ("Hyphenator") while it should be a list (e.g. __all__ = ["Hyphenator"]). https://stackoverflow.com/questions/44834/what-does-all-mean-in-python

In src\rinoh\highlight.py and src\rinoh\frontend\sphinx\nodes.py there are alerts on using-constant-test.
In both cases the function "warn" appears in the if statement but not as a call (e.g., warn()).
Later the function is called. What is the intention?
elif warn:
 warn('Could not lex literal_block as "%s". '
 'Highlighting skipped.' % language)

**Comment by evidencebp** [comment]
In src\rinoh\frontend\rst\nodes.py there is an alert on pointless statment.
Are properties being created on the fly and the self.title aimed to check the existence?
class Admonition(DocutilsGroupingNode):
 grouped_flowables_class = rt.Admonition

 def children_flowables(self, skip_first=0):
 try:
 self.title
 return super().children_flowables(skip_first=1)
 except AttributeError:
 return super().children_flowables()

**Comment by evidencebp** [comment]
In src\rinoh\element.py method build_document of class DocumentElement has an empty implementation of pass. 
It is implemented in subclasses. Might be better to enforce it by raising NotImplementedError.

**Comment by evidencebp** [comment]
In src\rinoh\stylesheets\matcher.py there is a wildcard import.
Code uses from rinoh.styleds import * 
This hides the imported objects and might lead to collisions in the future if objects of the same name are imported with wildcards.  styleds is a collection of imports, aimed to ease related imports. 
I copied the used imports from there and deleted the unused ones.  
This adds many imports and does not  benefit from styleds, so this is a downside of fixing this alert.

@brechtm, can you guide me regadding these cases?

**Comment by evidencebp** [comment]
Thank you very much for your detailed feedback!

The main guideline in this experiment is that the repository should benefit from them.
If you have time to review the other changes, that will be great.
Otherwise, merging the interventions that you like and fixing the bug will be great.

As for the benefit of linters, this is part of what the experiment tries to evaluate.
In general, it seems that complexity reducing alerts (like too-many-branches) do help to reduce tendency to bugs.
However, there are indeed many cases in which it desn't help (e.g., there were no bugs to begin with), so indeed it should not be applied blindly.

**Comment by brechtm** [comment]
Some more comments:
- `src/rinoh/attribute.py`: OK, but it seems arbitrary to extract part of the code to a method but not the code above it.
- `src/rinoh/image.py`: Looks good 👍
- `src/rinoh/font/opentype/cff.py`: I prefer the longer lines.
- `src/rinoh/backend/pdf/filter.py`: I wouldn't refactor the code like this. The `_handle_byte` function encapsulates only part of the code that forms a logical unit. For example, it would make more sense to still do the `byte != last_byte` check in _run_length_encoder_, but move the body of that branch to a function. This is similar to the refactoring in `src\rinoh\font\type1.py`, where a seemingly arbitrary part of the code is moved into a function just for the sake of fixing the linting error but ignoring any logical coherence.
- `src/rinoh/table.py`: Looks good 👍
- `src/rinoh/backend/pdf/xobject/purepng.py`: I assume this simply copied in the updated version of PurePNG? Note to self: I should include the PurePNG version number (at least in the commit message)
- `src/rinoh/flowable.py`: Looks good, except that I don't agree with putting comma's at the start of the line 😄
- `src/rinoh/hyphenator.py`: Looks good 👍
- `src/rinoh/paragraph.py` and `src/rinoh/backend/pdf/reader.py`: still to do...

In general, the _handle_*_ function names could be more descriptive.

**Comment by evidencebp** [comment]
src/rinoh/attribute.py: OK, but it seems arbitrary to extract part of the code to a method but not the code above it.
-- Yes, there can be other extraction.
Please remember that I'm not familiar with the project.

src/rinoh/flowable.py: Looks good, except that I don't agree with putting comma's at the start of the line 
-- A bit of a personal style, but I can change that.

src/rinoh/paragraph.py and src/rinoh/backend/pdf/reader.py: still to do...
-- Ok, go over them and then we will continue
